xml/hwbp_registers.xml

```diff
@@ -683,33 +683,42 @@
-        \begin{commentary} This trigger type is intended to be used as a single
-        step that's useful both for external debuggers and for software monitor
-        programs. For that case it is not necessary to support \FcsrIcountCount greater
-        than
-        1. The only two combinations of the mode bits that are useful in those
-        scenarios are \FcsrIcountU by itself, or \FcsrIcountM, \FcsrIcountS, and \FcsrIcountU all set.
-
-        If the hardware limits \FcsrIcountCount to 1, and changes mode bits instead of
-        decrementing \FcsrIcountCount, this register can be implemented with just 2
-        bits. One for \FcsrIcountU, and one for \FcsrIcountM and \FcsrIcountS tied together.  If
-        only the external debugger or only a software monitor needs to be
-        supported, a single bit is enough.
+        If \FcsrIcountCount is not 0, then every instruction completed or
+        trap taken from a mode where the trigger is enabled decrements
+        \FcsrIcountCount by 1, unless \FcsrIcountAction is 0, the current
+        mode is M-mode, and M-mode breakpoints are disabled via one of the
+        mechanisms described in Section~\ref{sec:mmtrigger}. When
+        \FcsrIcountCount is decremented from 1 to 0 then \FcsrIcountPending
+        becomes set. When \FcsrIcountPending is set, the trigger fires just
+        before any further instructions are executed in a mode where the
+        trigger is enabled. As the trigger fires, \FcsrIcountPending is
+        cleared.
+
+        \begin{commentary}
+        The intent of this design is to cleanly handle the case where
+        \FcsrIcountAction is 0, \FcsrIcountM is 0, \FcsrIcountU is 1,
+        \FcsrIcountCount is 1, and the U-mode instruction being executed
+        causes a trap into M-mode. In that case we want the entire M-mode
+        handler to be executed, and the debug trap to be taken before the
+        next U-mode instruction.
+        \end{commentary}
+
+        \begin{commentary}
+        This trigger type is intended to be used as a single step that's
+        useful both for external debuggers and for software monitor programs.
+        For that case it is not necessary to support \FcsrIcountCount greater
+        than 1.
         \end{commentary}
 
         <field name="type" bits="XLEN-1:XLEN-4" access="R" reset="3" />
         <field name="dmode" bits="XLEN-5" access="WARL" reset="0" />
         <field name="0" bits="XLEN-6:27" access="R" reset="0" />
 
         <field name="vs" bits="26" access="WARL" reset="0">
-          When set, enable this trigger in VS-mode.
-          This bit is hard-wired to 0 if the hart does not support
-          virtualization mode.
+            When set, enable this trigger in VS-mode.
         </field>
         <field name="vu" bits="25" access="WARL" reset="0">
-          When set, enable this trigger in VU-mode.
-          This bit is hard-wired to 0 if the hart does not support
-          virtualization mode.
```

Comment:
Removing this will make these bits consistent with `u` and `s` which are also optional modes.  However, the spec is not consistent across registers.  For instance, `dcsr` says that `ebreaku` is hardwired to 0 if U mode isn't supported but none of the trigger registers say the same about the `u` bit.  All of the `vs` and `vu` bits in the other registers are inconsistent with `s` and `u` in those same registers (which is my fault) and this change only fixes one.

My preference would be to leave these sentences in and add similar sentences to all the existing `u` and `s` bits so there is no confusion about legal values on these WARL fields.

Reply:
I would like not to have to mention for every feature that if the feature doesn't exist, the bits are wired to 0. That's the default for (I think) every bit we define anywhere in this spec. We already have some language that talks about the fact that all triggers are WARL at https://github.com/riscv/riscv-debug-spec/blob/master/trigger.tex#L176. Is there a place where we can make a general statement, instead of replacing the same boilerplate over and over?

I guess I don't feel super strongly about this. It just feels like we're repeatedly stating the obvious. But maybe that's better than for somebody (who likely did not read the entire document) to be unsure.

Reply:
The privileged spec explicitly describes what happens for each bit if it's affected by a lack of S-mode or U-mode, etc.  Personally, I never take for granted that people will have enough understanding to do what I think is obvious.  On the other hand, what I think is obvious is sometimes not the correct thing to do because I am the one who lacks understanding.  Making things explicit helps get everyone "on the same page" at the expense of making the spec have an extra page or two.

Reply:
I'm making it explicit.